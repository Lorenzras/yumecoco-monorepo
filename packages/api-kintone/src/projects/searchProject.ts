import { ktRecord } from './../client';
import { appId, RecordKey, RecordType } from './config';

export const searchProjects = async (search: string) => {

  const fieldProjName: RecordKey = 'projName';

  return (await ktRecord()).getRecords({
    app: appId,
    query: `${fieldProjName} like "${search}"`,
    totalCount: true,
  })
    .then(r => r.records as unknown as RecordType[]);
};