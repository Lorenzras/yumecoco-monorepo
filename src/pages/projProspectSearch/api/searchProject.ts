import { APPIDS, KintoneRecord } from '../../../api/kintone';
import { TypeOfForm } from '../form';
import { simplifyKintoneRecords } from './simplifyKintoneRecords';

// Unpack a promise
type Unpack<T> = T extends Promise<infer U> ? U : T;
export type TSearchResult =  Unpack<ReturnType<typeof searchProject>>;
export type TKeyOfSearchResult = keyof TSearchResult[number];

type KeyOfAgents = keyof TypeOfProjectDetails['agents']['value'][number]['value'];
type KeyOfCustGroup = keyof TypeOfProjectDetails['custGroup']['value'][number]['value'];
type KeyOfFlatProjDetails = (KeyOfProjectDetails | KeyOfAgents | KeyOfCustGroup);

const getKeyOfProj = (k: KeyOfFlatProjDetails ) => k;

const mainSearchCondition = (mainSearch?: string) => {
  if (!mainSearch) return undefined;

  const fieldsToCompare : KeyOfFlatProjDetails[]  = [
    '$id', 'projName', 'custGroupId', 'memo', 'agentName',
    'custName', 'custNameReading', 'rank',
  ];

  const exactFields : KeyOfFlatProjDetails[] = ['$id', 'custGroupId'];

  const condition = fieldsToCompare
    .map(f => `(${f} ${exactFields.includes(f) ? '=' : 'like'} "${mainSearch}")`)
    .join(' or ');
  return `(${condition})`;
};

const rankCondition = (rank?: TProjRank[]) => {
  if (!rank?.length) return;

  const fieldName: KeyOfFlatProjDetails = 'rank';
  const condition = rank.map(r => {
    return `${fieldName} = "${r}"`;
  }).join(' or ');

  return `(${condition})`;
};



export const searchProject = async (form : Partial<TypeOfForm>) => {
  const {
    mainSearch, rank, custGroupId, projId, projName,

    schedContractPriceMin, schedContractPriceMax,
    schedContractDateMin, schedContractDateMax,
    planApplicationDateMin, planApplicationDateMax,
    estatePurchaseDateMin, estatePurchaseDateMax,

    memo,
  } = form;
  const fields : KeyOfProjectDetails[] = [
    'projName', '$id', 'custGroupId', 'memo',
    'agents', 'custGroupAgents',
    'custGroup', 'rank',
    'store', '更新日時', '作成日時',
    'schedContractDate', 'estatePurchaseDate', 'planApplicationDate',
    'schedContractPrice',
  ];


  const allConditions = [
    mainSearchCondition(mainSearch),
    rankCondition(rank),
    custGroupId ? `(${getKeyOfProj('custGroupId')} = "${custGroupId}")` : undefined,
    projId ? `(${getKeyOfProj('$id')} = "${projId}")` : undefined,
    projName ? `(${getKeyOfProj('projName')} like "${projName}")` : undefined,
    schedContractPriceMin?.toString() ? `(${getKeyOfProj('schedContractPrice')} >= ${schedContractPriceMin} and ${getKeyOfProj('schedContractPrice')} != "")` : undefined,
    schedContractPriceMax?.toString() ? `(${getKeyOfProj('schedContractPrice')} <= ${schedContractPriceMax})` : undefined,

    schedContractDateMin ? `(${getKeyOfProj('schedContractDate')} >= "${schedContractDateMin}")` : undefined,
    schedContractDateMax ? `(${getKeyOfProj('schedContractDate')} <= "${schedContractDateMax}")` : undefined,

    planApplicationDateMin ? `(${getKeyOfProj('planApplicationDate')} >= "${planApplicationDateMin}")` : undefined,
    planApplicationDateMax ? `(${getKeyOfProj('planApplicationDate')} <= "${planApplicationDateMax}")` : undefined,

    estatePurchaseDateMin ? `(${getKeyOfProj('estatePurchaseDate')} >= "${estatePurchaseDateMin}")` : undefined,
    estatePurchaseDateMax ? `(${getKeyOfProj('estatePurchaseDate')} <= "${estatePurchaseDateMax}")` : undefined,

    memo ? `(${getKeyOfProj('memo')} like "${memo}")` : undefined,

  ]
    .filter(Boolean)
    .join(' and ');

  const result = await KintoneRecord.getAllRecords({
    app: APPIDS.constructionDetails,
    condition: allConditions,
    fields: fields,
  });

  return simplifyKintoneRecords(result as unknown as TypeOfProjectDetails[]);
};

export const initialSearch = async () => {

  const orderBy: KeyOfProjectDetails = '更新日時';

  const result = await KintoneRecord.getRecords({
    app: APPIDS.constructionDetails,
    query: `order by ${orderBy} desc limit ${10}`,
  });


  return simplifyKintoneRecords(result.records as unknown as TypeOfProjectDetails[]);
};