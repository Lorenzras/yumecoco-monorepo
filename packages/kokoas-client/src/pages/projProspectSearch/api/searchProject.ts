import { KFlatProjects } from './../../../../../types/src/common/projects';
import { AppIds } from 'config';
import { KProjects, IProjects, TProjRank } from 'types';
import { KintoneRecord } from '../../../api/kintone';
import { TypeOfForm } from '../form';
import { simplifyKintoneRecords } from './simplifyKintoneRecords';

// Unpack a promise
type Unpack<T> = T extends Promise<infer U> ? U : T;
export type TSearchResult =  Unpack<ReturnType<typeof searchProject>>;
export type TKeyOfSearchResult = keyof TSearchResult[number];


const getKeyOfProj = (k: KFlatProjects ) => k;

const mainSearchCondition = (mainSearch?: string) => {
  if (!mainSearch) return undefined;

  const fieldsToCompare : KFlatProjects[]  = [
    '$id', 'projName', 'custGroupId', 'memo', 'agentName',
    'custName', 'custNameReading', 'rank',
  ];

  const exactFields : KFlatProjects[] = ['$id', 'custGroupId'];

  const condition = fieldsToCompare
    .map(f => `(${f} ${exactFields.includes(f) ? '=' : 'like'} "${mainSearch}")`)
    .join(' or ');
  return `(${condition})`;
};

const rankCondition = (rank?: TProjRank[]) => {
  if (!rank?.length) return;

  const fieldName: KFlatProjects = 'rank';
  const condition = rank.map(r => {
    return `${fieldName} = "${r}"`;
  }).join(' or ');

  return `(${condition})`;
};



export const searchProject = async (form : Partial<TypeOfForm>) => {
  const {
    mainSearch, rank, custGroupId, projId, projName,

    schedContractPriceMin, schedContractPriceMax,
    schedContractDateMin, schedContractDateMax,
    planApplicationDateMin, planApplicationDateMax,
    estatePurchaseDateMin, estatePurchaseDateMax,

    memo,
  } = form;
  const fields : KProjects[] = [
    'projName', '$id', 'custGroupId', 'memo',
    'agents', 'custGroupAgents',
    'custGroup', 'rank',
    'store', '更新日時', '作成日時',
    'schedContractDate', 'estatePurchaseDate', 'planApplicationDate',
    'schedContractPrice',
  ];


  const allConditions = [
    mainSearchCondition(mainSearch),
    rankCondition(rank),
    custGroupId ? `(${getKeyOfProj('custGroupId')} = "${custGroupId}")` : undefined,
    projId ? `(${getKeyOfProj('$id')} = "${projId}")` : undefined,
    projName ? `(${getKeyOfProj('projName')} like "${projName}")` : undefined,
    schedContractPriceMin?.toString() ? `(${getKeyOfProj('schedContractPrice')} >= ${schedContractPriceMin} and ${getKeyOfProj('schedContractPrice')} != "")` : undefined,
    schedContractPriceMax?.toString() ? `(${getKeyOfProj('schedContractPrice')} <= ${schedContractPriceMax})` : undefined,

    schedContractDateMin ? `(${getKeyOfProj('schedContractDate')} >= "${schedContractDateMin}")` : undefined,
    schedContractDateMax ? `(${getKeyOfProj('schedContractDate')} <= "${schedContractDateMax}")` : undefined,

    planApplicationDateMin ? `(${getKeyOfProj('planApplicationDate')} >= "${planApplicationDateMin}")` : undefined,
    planApplicationDateMax ? `(${getKeyOfProj('planApplicationDate')} <= "${planApplicationDateMax}")` : undefined,

    estatePurchaseDateMin ? `(${getKeyOfProj('estatePurchaseDate')} >= "${estatePurchaseDateMin}")` : undefined,
    estatePurchaseDateMax ? `(${getKeyOfProj('estatePurchaseDate')} <= "${estatePurchaseDateMax}")` : undefined,

    memo ? `(${getKeyOfProj('memo')} like "${memo}")` : undefined,

  ]
    .filter(Boolean)
    .join(' and ');

  const result = await KintoneRecord.getAllRecords({
    app: AppIds.projects,
    condition: allConditions,
    fields: fields,
  });

  return simplifyKintoneRecords(result as unknown as IProjects[]);
};

export const initialSearch = async () => {

  const orderBy: KProjects = '更新日時';

  const result = await KintoneRecord.getRecords({
    app: AppIds.projects,
    query: `order by ${orderBy} desc limit ${10}`,
  });


  return simplifyKintoneRecords(result.records as unknown as IProjects[]);
};