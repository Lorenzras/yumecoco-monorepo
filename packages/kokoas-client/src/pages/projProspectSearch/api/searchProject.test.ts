import { searchProject } from './searchProject';

describe('Prospect Search', ()=>{
  it('prospect', async ()=>{
    const result = await searchProject({
      mainSearch: 'Lorenz',

    });

    console.log(result.length);

    expect(result).toMatchSnapshot();
  }, 30000);
});