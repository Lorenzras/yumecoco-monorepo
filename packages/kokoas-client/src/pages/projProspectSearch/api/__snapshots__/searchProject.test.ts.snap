// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prospect Search prospect 1`] = `
Array [
  Object {
    "$id": "103",
    "agents": Array [
      Object {
        "agentType": "cocoConst",
        "employeeName": "坂神　隆仁",
      },
      Object {
        "agentType": "cocoConst",
        "employeeName": "",
      },
    ],
    "custGroup": Array [],
    "custGroupId": "164",
    "memo": "",
    "projName": "Lorenz Ras様邸 設備交換工事",
  },
  Object {
    "$id": "104",
    "agents": Array [
      Object {
        "agentType": "cocoConst",
        "employeeName": "黒木　麻衣子",
      },
      Object {
        "agentType": "cocoConst",
        "employeeName": "",
      },
    ],
    "custGroup": Array [],
    "custGroupId": "164",
    "memo": "ccvxcxcxc",
    "projName": "Lorenz Ras様邸 蓄電池",
  },
  Object {
    "$id": "109",
    "agents": Array [
      Object {
        "agentType": "cocoConst",
        "employeeName": "沖永　誠治",
      },
      Object {
        "agentType": "cocoConst",
        "employeeName": "大井　道晴",
      },
    ],
    "custGroup": Array [],
    "custGroupId": "178",
    "memo": "",
    "projName": "LORENZ RAS様邸 営繕工事",
  },
  Object {
    "$id": "110",
    "agents": Array [
      Object {
        "agentType": "cocoConst",
        "employeeName": "大井　道晴",
      },
      Object {
        "agentType": "cocoConst",
        "employeeName": "大井　道晴",
      },
    ],
    "custGroup": Array [],
    "custGroupId": "178",
    "memo": "",
    "projName": "LORENZ RAS様邸 その他工事",
  },
]
`;
